### PR TITLE
fix payment methods loading on multi-distributor order cycle checkout page

### LINKED ISSUE #9873: Multi-distributor order cycle: Available shipping/payment methods don't respect OC settings (step 4 - checkout options)
## Description

While testing #9755 it was observed that for **multiple-distributor** order cycles, the checkout options regarding payment and shipping methods are not applied correctly. Instead of displaying only the selected payment/shipping methods for the different distributors, all activated methods are displayed to the customers.
Example: 
- Distributor A offers payment method A and shipping method A.
- Distributor B offers payment method B and shipping method B.
- Result: On checkout payment methods A and B as well as shipping methods A and B are available at both distributors A and B. --> wrong

## Expected Behavior

On checkout only the in step 4 of the order cycle creation selected shipping/payment methods should be available for the respective distributor.

## Actual Behaviour

On checkout all of the in step 4 of the order cycle creation selected shipping/payment methods are available at all distributors.

## Steps to Reproduce

1. Set up two hubs (A and B) (setting 'sells = any')
2. Set up two payment methods (A and B) and apply them to both hubs
3. Set up two shipping methods (A and B) and apply them to both hubs
4. For one of the hubs create an order cycle and on step 3 add both hubs as distributors
5. On step 4 activate shipping method A and payment method A for distributor A
6. On step 4 activate shipping method B and payment method B for distributor B
7. As a customer visit one of the shopfronts, add products and proceed to checkout
8. Observe that you will be able to select both shipping methods (A and B) and payment methods (A and B) for checkout

## Animated Gif/Screenshot

It seems like the checkout isn't considering which shop I am actually shopping from. Instead the combination of available shipping methods is offered for selection. Some examples below, shopping at 'Konrad Non-Producer Hub', so only those options should be offered:

image
image
image

The same is true for **payment methods**.

## Workaround

None.

## Severity
bug-s2: a non-critical feature is broken, no workaround

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by abdellani**
#### What? Why?

- fix partially #9873

#### What should we test?

- described on #9873

#### Release notes

Changelog Category: Technical changes

### CODE REVIEW

**Review by dacook — CHANGES_REQUESTED** [review]
This looks like a good fix, however it's not covered by a spec to prove it. 

Can you please add a spec to cover this condition? I find it surprisingly satisfying and helpful to follow this order:
1. build the failing spec to show the bug (**red**),
2. apply the fix to see it pass (**green**).

I notice that this code requires up to two DB queries (first to build a list of IDs, then another to get the filtered payment methods). After adding the spec, perhaps we can try to

3. update it to use only one query (**refactor**)

File: app/services/order_available_payment_methods.rb [hunk 1]
```diff
@@ -30,8 +30,13 @@ def payment_methods_before_tag_rules_applied
...
       distributor.payment_methods.where(
-        id: order_cycle.distributor_payment_methods.select(:payment_method_id)
+        id: available_distributor_payment_methods_ids
       )
     end.available.select(&:configured?)
   end
+
+  def available_distributor_payment_methods_ids
+    order_cycle.distributor_payment_methods.where(distributor_id: distributor.id)
+      .select(:payment_method_id)
```
Comment (jibees) [inline app/services/order_available_payment_methods.rb:40]:
```suggestion
    order_cycle.distributor_payment_methods
      .where(distributor_id: distributor.id)
      .select(:payment_method_id)
```

**Review by jibees — APPROVED** [review]
Very close to https://github.com/openfoodfoundation/openfoodnetwork/pull/10076, so I'll be approve this one too!

**Comment by dacook** [comment]
> For the optimization, I'm not sure how can I do it in one query. Can you please guide me?

Sorry I didn't get back to you this time, Mohamed. Thanks for pushing this forward anyway.

**Comment by filipefurtad0** [comment]
Continuing the testing notes from here:

Checking out on Shop A (/payments)

image

Checking out on Shop B (/payments)

image

We can see that only the payment method enabled for a given shop is appearing on checkout, for the customer.

Merging!

**Pull request merged**